app/Http/Controllers/SupplyController.php

```diff
@@ -0,0 +1,269 @@
+                $userdata = (array)json_decode(file_get_contents("{$aduser_endpoint}/get/{$impressionId}"), true);
+            } else {
+                $userdata = [];
+            }
+        }
+
+        $keywords = array_merge($context, $userdata);
+
+        $banners = BannerFinder::getBestBanners($zones, $keywords);
+
+        foreach ($banners as &$banner) {
+            if ($banner) {
+                $banner['pay_to'] = Esc::normalizeAddress(config('app.adshares_address'));
+            }
+        }
+
+        $response->setContent(json_encode($banners, JSON_PRETTY_PRINT));
+        return $response;
+    }
+
+    public function findScript(Request $request)
+    {
+        $params = [json_encode($request->getSchemeAndHttpHost()), json_encode(config('app.aduser_endpoint'))];
+
+        $jsPath = config('app.env') == 'production' ? resource_path('assets/js/tmp-copy/find.x.js') : resource_path('assets/js/tmp-copy/find.min.js');
+
+        $response = new StreamedResponse();
+        $response->setCallback(function () use ($jsPath, $request, $params) {
+            echo str_replace([
+                "'{{ ORIGIN }}'",
+                "'{{ ADUSER }}'",
+            ], $params, file_get_contents($jsPath));
+        });
+
+        $response->headers->set('Content-Type', 'text/javascript');
+
+        $response->setCache(array(
+            'etag' => md5(md5_file($jsPath) . implode(':', $params)),
+            'last_modified' => new \DateTime('@' . filemtime($jsPath)),
+            'max_age' => 3600 * 24 * 30,
+            's_maxage' => 3600 * 24 * 30,
+            'private' => false,
+            'public' => true
+        ));
+
+        if (! $response->isNotModified($request)) {
+            // TODO: ask Jacek
+        }
+        return $response;
+    }
+
+    public function logNetworkClick(Request $request, $id)
+    {
+        if ($request->query->get('r')) {
+            $url = Utils::UrlSafeBase64Decode($request->query->get('r'));
+            $request->query->remove('r');
+        } else {
+            $banner = NetworkCampaign::getRepository($this->getDoctrine()->getManager())->findOneBy([
+                'uuid' => $id
+            ]);
+
+            if (! $banner) {
+                throw new NotFoundHttpException();
+            }
+
+            $url = $banner->getClickUrl();
+        }
+        $qString = http_build_query($request->query->all());
+        if ($qString) {
+            $qPos = strpos($url, '?');
+
+            if ($qPos === false) {
+                $url .= '?' . $qString;
+            } elseif ($qPos == strlen($url) - 1) {
+                $url .= $qString;
+            } else {
+                $url .= '&' . $qString;
+            }
+        }
+
+        $logIp = bin2hex(inet_pton($request->getClientIp()));
+
+
+        $cid = Utils::getRawTrackingId($request->query->get('cid'));
+        $tid = Utils::getRawTrackingId($request->cookies->get('tid')) ?: $logIp;
+        // TODO get user / website / zone
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1135)